dungeon generation optimization

ericp275 opened the pull request:
I played around with Tiff's code because it was taking a long time on my computer. I haven't tested this new code thoroughly so there might be bugs in it, but it seems to be working pretty well.

Basically instead of building all the walls and destroying them I just keep track of where the walls are going to be and build them at the end. Also I put the spawning objects code before the removeLagWalls code so that way objects don't spawn in areas the player can't reach. There's a lot less object creation now so the scene loading time should be faster.

I also tried to fix the bug with the hit sound effect. I can't really tell if its working or not.

yhbautista commented on fiscal-shock/Assets/Scripts/Graphs/Voronoi.cs:
```diff
@@ -18,6 +18,7 @@ public class Voronoi : Graph {
         /// Polygons representing faces of the Voronoi diagram.
         /// </summary>
         public List<Cell> cells { get; }
+        public List<Edge> allEdges => cells.SelectMany(c => c.sides).Distinct().ToList();
```
See the Graph code. Please.

ericp275 replied:
I didn't realize that was already there.
Thank you.

yhbautista commented on fiscal-shock/Assets/Scripts/ProceduralGeneration/ObjectGeneration/Walls.cs:
```diff
@@ -136,106 +139,134 @@ public static class Walls {
-                        shortDestroyedWalls.Add(er);
+            List<Edge> shortDestroyedWalls = new List<Edge>();
+            foreach( Edge vEdge in d.vd.allEdges){
+                float minDistanceToTree = 100000;
```
Where are these numbers coming from? Where are your comments? While I'm quite the hypocrite for saying this, I do need this code to be readable if this goes through for the Voronoi diagram, especially since I am _actively working in this code_.

yhbautista commented on fiscal-shock/Assets/Scripts/ProceduralGeneration/ObjectGeneration/Walls.cs:
```diff
@@ -136,106 +139,134 @@ public static class Walls {
+                float minDistanceToTree = 100000;
+                foreach (Edge e in d.spanningTree) {
+                    float slope = (e.p.y - e.q.y) / (e.p.x - e.q.x);
+                    float intercept = e.p.y - (slope * e.p.x);
+                    float dist = 100000;
+                    float candidate0 = vEdge.p.y - ((slope * vEdge.p.x) + intercept);
+                    if(Mathf.Abs(candidate0) < dist){
+                        dist = Mathf.Abs(candidate0);
+                    }
+                    float candidate1 = vEdge.p.x - ((vEdge.p.x - slope) / intercept);
+                    if(Mathf.Abs(candidate1) < dist){
+                        dist = Mathf.Abs(candidate1);
+                    }
+                    float candidate2 = vEdge.q.x - ((vEdge.q.x - slope) / intercept);
+                    if(Mathf.Abs(candidate2) < dist){
+                        dist = Mathf.Abs(candidate2);
+                    }
+                    float candidate3 = vEdge.q.y - ((slope * vEdge.q.x) + intercept);
+                    if(Mathf.Abs(candidate3) < dist){
+                        dist = Mathf.Abs(candidate3);
+                    }
+                    if((candidate0 > 0 && candidate3 < 0) || (candidate0 < 0 && candidate3 > 0)){
+                        dist = 0;
+                    }
+                    if(dist < minDistanceToTree){
+                        minDistanceToTree = dist;
+                    }
+                    if(minDistanceToTree < 5){
+                        break;
                     }
-                    hit.collider.gameObject.name = $"Destroyed {hit.collider.gameObject.name}";
-                    UnityEngine.Object.Destroy(hit.collider.gameObject);
                 }
```
While you understand your math (now), anyone that attempts to look at your code will not know what your motive is. Unless practice is to not comment code (which is not the case for us), please leave explanations as to why you're using a specific technique and what it's doing.

yhbautista commented:
For future reference, you will piss off _a lot_ of people by working on code and not informing the team. You played around with code that is vital to the enemy movement, and there are no comments to explain what is going on. I don't have time to try to understand the math. I, and just about every other person on the team, am on a full time schedule. So, unless you want to really piss off a future team member (this is me with four hours of sleep and trying to hold in my temper, if you thought this was bad), please inform the team of your motives before a PR is put through, especially when there's a chance another team member might be working on it. People really don't like to resolve last minute merge conflicts, and this PR is reeking of them. At least a head's up would've been nice.

zalfenior commented on fiscal-shock/Assets/Scripts/AI/EnemyHealth.cs:
```diff
@@ -105,6 +106,9 @@ public class EnemyHealth : MonoBehaviour {
         if (enmityCounter >= maxEnmityDuration) {
             enmityActive = false;
         }
+        if(timeSinceHit < 2){
```
Whats the purpose of this specifically?

ericp275 replied:
It measures how long ago the last hit was. So if it has been long enough the sound can play again.

zalfenior commented on fiscal-shock/Assets/Scripts/ProceduralGeneration/ObjectGeneration/Walls.cs:
```diff
@@ -17,8 +18,11 @@ public static class Walls {
+        }
+
+        public static void buildWalls(Dungeoneer d){
+            buildWallsToKeep(d);
```
Does this make it run any better?

zalfenior commented on fiscal-shock/Assets/Scripts/ProceduralGeneration/ObjectGeneration/Walls.cs:
```diff
@@ -136,106 +139,134 @@ public static class Walls {
-                Vector3 p = e.p.toVector3AtHeight(1);
-                Vector3 q = e.q.toVector3AtHeight(1);
-                Vector3 direction = (q-p).normalized;
-                // uncomment to debug raycasts
```
I know a lot of the code with vectors was being used by others in their ongoing branches. Is there a place this was readded or was it just removed?

ericp275 replied:
I think these vectors were just used to do a raycast and find the walls intersecting hallways so they could be removed.

zalfenior commented on fiscal-shock/Assets/Scripts/ProceduralGeneration/ObjectGeneration/Walls.cs:
```diff
@@ -136,106 +139,134 @@ public static class Walls {
+                    if(dist < minDistanceToTree){
+                        minDistanceToTree = dist;
+                    }
+                    if(minDistanceToTree < 5){
```
Is this to generate the walls specifically? I'm just guessing here because there isn't any documentation. Also, I know a lot of this interferes with other peoples code so this might not have been the best thing to mess with

zalfenior commented on fiscal-shock/Assets/Scripts/ProceduralGeneration/ObjectGeneration/Walls.cs:
```diff
@@ -89,9 +93,10 @@ public static class Walls {
         /// <param name="d"></param>
         /// <param name="p"></param>
-        private static void constructWallsOnPolygon(Dungeoneer d, Polygon p) {
+        private static void constructWallsOnPolygon(Polygon p) {
```
Why was dungeoneer removed here?

ericp275 replied:
It was used to build walls. But this code doesn't build the walls yet. It just marks the walls as going to be built.

yhbautista commented:
> For future reference, you will piss off _a lot_ of people by working on code and not informing the team. You played around with code that is vital to the enemy movement, and there are no comments to explain what is going on. I don't have time to try to understand the math. I, and just about every other person on the team, am on a full time schedule. So, unless you want to really piss off a future team member (this is me with four hours of sleep and trying to hold in my temper, if you thought this was bad), please inform the team of your motives before a PR is put through, especially when there's a chance another team member might be working on it. People really don't like to resolve last minute merge conflicts, and this PR is reeking of them. At least a head's up would've been nice.

In follow up to Zach's comment (and now that I've calmed down some), I would like to say that this is not an unappreciated thought. The code could use some optimization. But Zach is right. It's far too late in the semester to work on a drastic change like this. If we wanted to make a change like this, it would have to be thoroughly tested by just about every member on the team. My comment from before still stands, but I'm sorry if I came on too aggressive in my last comment.

pagewin commented:
I like the idea of improving performance and I know I had to choose a lazy route to get walls working in exchange for longer load times. The same thing for getting objects generated everywhere. But I want to echo the concern about a change like this that was not thoroughly tested and greatly affects the pathfinding. There are certain traits in the graph/dungeon (mainly with respect to walls) that Yadira is expecting while writing the pathfinding code and taking care of all of the edge cases she mentioned at the meeting.

I can sort of see the intent in the code after looking through it. Unfortunately, in reality, it doesn't work the way it should. "Runs without crashing" doesn't mean it's working right. The code that dealt with determining how wide halls are appears to be completely gone.
![1587665025](https://user-images.githubusercontent.com/20141682/80134397-3b011b80-858e-11ea-89c3-106205d783de.png)
The halls here are way too narrow. Looking at the blue graph, I'm not even sure Yadira's pathfinding code could make it through here. I think she might be removing edges that are connected only to walls, which is the case for these ones in the hall. Not that it matters, because there happened to be an object rotated just the right way to block passage even for the player. The temple is intended to have wide corridors and this change removes that.
![1587664396](https://user-images.githubusercontent.com/20141682/80134590-7996d600-858e-11ea-9014-ecbf6e86318f.png)
The walls extending off into the distance aren't a big deal, but they weren't being generated originally. The edges they're on do exist on the graph.

But I'm really concerned about how you tested this. I don't know if you tested the mines at all or didn't look at what the mines was like before.
![1587665294](https://user-images.githubusercontent.com/20141682/80135157-5b7da580-858f-11ea-87a1-21411b869917.png)
This is *not* what the mines are supposed to look like. It should be small, claustrophobic corridors and small rooms. Instead, it's wide open (nice cave-ins there). This change trashes the intended characteristics of the mines.

I'm pretty sure this is a relic of using the spanning tree and determining where to put walls. The mines has a much more dense spanning tree, so I think your code was way too conservative in deciding where to put walls. But it's too aggressive and makes extremely skinny corridors in the temple.

yhbautista commented:
To expand on the issues that Tiff gave:
- The colliders of different enemies are of different sizes. We can't expect every enemy to fit inside those narrow corridors--and they won't. If you look at line 164 from the Walls script, you'll see that the code to account for the largest character controller was removed (currently set to 6 for the blaster's character controller radius, as seen somewhere further up in that script). 
- Because the corridors are so narrow, I now no longer have the guarantee of at least one connecting edge of the graph, which was a guarantee when the the fattest controller was still being accounted for. I have edge case code dealing with that, but even that edge case code could be failing its own edge case. See where this goes? Pathfinding on the same graph where you place obstacles is already extremely vulnerable code, and having this change go through so late in the semester will mean that all of the guarantees that I had with the Poisson and the fattest character controller no longer exist. I need those guarantees to exist this late into the project.
- With the hallways so narrow AND the hallways only separated by one wall, a major edge case in the pathfinding completely breaks. I had to account for instances where two rooms or two hallways were only separated by a single wall and make sure that that single wall was not counted among the navigable Voronoi edges. In those cases, I removed navigable edges that were walls and had more than one non-walled in cell associated with it. While I could find another solution to this problem, it's too late to try something else, especially if there's no way to adjust how narrow or wide the hallways are.

ericp275 commented:
So I added some more comments to help explain the code. I changed the hall width to a variable. Playing around with it might make the dungeons look more like there supposed to. I didn't know how to view the mine from above so I just guessed that it looked right based on the in-game view. Obviously I was off.

I was well aware that this was a major change. I started experimenting with it last night after the meeting, and this morning when I got to a point where it seemed to work, I posted it for you guys to look at. I know it is late in the semester and that is exactly why I put it up as quickly as possible (before I had completely documented it), so you guys would have time to look at it.

I tried to not change the code in anyway that would effect path-finding. Obviously I don't know all the problems that could be created and that's exactly why I posted it here for the team members to look at. If it does create problems we can just not merge the branch (but we can still keep the record of the code for possible later optimization). I apologize for changing the code without telling anyone. I thought it was fine since I kept the changes on a separate branch.

yhbautista commented:
This is how I feel about the situation. These are my feelings on the matter, and it's up to you to figure out whether or not you think they are valid. I'll try my best, though, to give an explanation as to why I felt the way I did.
- I had already stated multiple times during meetings, and it was even a discussion on discord, that I would be using the Voronoi diagram for pathfinding. I had hoped that anyone looking at this code this late in the semester would understand the implications of using the same graph to generate halls and rooms would have on navigating that graph. It's understandable, though, if it's something you missed. Tiff is the only person on this team that 100% understands the code that she wrote, so it's easy to miss the implications of how merging these different components would cause problems.
- In a team at a job, your team, for the most part, is going to need to know what code you're messing with. That's why task and project trackers exist. It's extremely impolite to mess around in code that someone else is working on when there is no real need. As Tiff said, this is a major, developing feature. I was hoping that Tiff's changes to the code would be the last major changes to it, and to see this PR this morning took me by surprise. Additionally, when people make major changes to code, it's polite to resolve the merge conflicts on your side instead of placing the responsibility on someone else (in defense of Tiff, who did this every time she made a major change to the code of any member of the team). If this PR goes through, then the merge conflict resolution falls on me. I was willing to take it with Tiff's code because it would improve a lot of major aspects of the navigation that I needed, and I was expecting it for weeks before the PR went through, but this PR would be a major headache to resolve on my side, and would required major changes to code that I've almost completed working out the major kinks on. Additionally, I wasn't expecting it, as I said before.
- Let me try to give an equivalent scenario. Imagine you had been working hard on that enemy weapons code, then I decided to completely rewrite whatever code existed in master without telling you. I only knew that you were vaguely working on that code, but I knew for a fact that you were. I put through the PR anyway because I think it'll significantly improve a feature of the game, but that PR going through would require that you make major changes to code that you were almost done working on. How would you feel? I'm not saying my anger was valid, and I could've used more patience or waited until later in the day to respond, and I'm sorry for that. At the same time, though, I believe that I am justified in being annoyed and frustrated with this PR. 
I hope that some of this at least reflects a bit of what I felt during the course of this afternoon. Like I said, you can choose whether or not your believe that my feelings were valid or not, but at least I've explained my side of the matter.

pagewin commented on fiscal-shock/Assets/Scripts/ProceduralGeneration/ObjectGeneration/Walls.cs:
```diff
@@ -136,106 +139,134 @@ public static class Walls {
+                    if(!cEdge.isWall){
+                        openCell = true;
+                    }
+                    if(cEdge.p.x < 0 || cEdge.p.x > 500 || cEdge.p.y < 0 || cEdge.p.y > 500 || cEdge.q.x < 0 || cEdge.q.x > 500 || cEdge.q.y < 0 || cEdge.q.y > 500){
```
What does the 500 mean here? You should be getting the bounds from `Dungeoneer.currentDungeonType` and not assuming 500 is the min/max, since they can be configured differently per dungeon type.

pagewin commented on fiscal-shock/Assets/Scripts/ProceduralGeneration/ObjectGeneration/Walls.cs:
```diff
@@ -136,106 +139,134 @@ public static class Walls {
+
+            //For each edge in the voronoi graph, chech if ether vertex is within the Hall Width distance to the spaning tree. If it is, remove the wall.
+            foreach( Edge vEdge in d.vd.edges){
+                float hallWidth = 5;
```
Hall width is set in the `DungeonType` as it's configurable per-dungeon, so this should be checking `d.currentDungeonType.hallWidth` (or whatever the variables are named)

zalfenior commented:
So, I think we just have to close this pull request. There's just not enough time in the semester to adjust to these changes and change A* etc around to accommodate. I think it would be best for us to mark this as a potential alternative way of making the generator work. We can easily get a paragraph or two in the final report about it. 

On a positive note, I do like how some of the maps showed up, they could have easily been new level types with some adjustments. If we decide to keep working on this project I could see that being implemented at the very least.
